### PR TITLE
breaking(IReactContext): Extract IReactContext from ReactContext, make ThemedReactContext pure delegation

### PR DESCRIPTION
**Pull request by matthargett**

### CODE REVIEW

**Comment by rozele** [comment]
I think we need to go more minimalist on this, and also phase out the diffs. We can pretty easily add the `IReactContext` and massage the base classes so they aren't breaking for almost any modules or view managers.

File: ReactWindows/ReactNative.Shared/Bridge/JavaScriptModuleBase.Generated.cs [hunk 1]
```diff
@@ -124,6 +124,5 @@ protected void Invoke(object arg0, object arg1, object arg2, object arg3, [Calle
         {
             Invoke(new[] { arg0, arg1, arg2, arg3 }, caller);
         }
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/JavaScriptModuleBase.Generated.cs]: Going to mark a few of these things for clean up in a separate PR. This is one of them.

File: ReactWindows/ReactNative.Shared/Bridge/ReactDelegateFactoryBase.cs [hunk 2]
```diff
@@ -125,7 +125,7 @@ protected static IPromise CreatePromise(JToken resolveToken, JToken rejectToken,
 
         class Callback : ICallback
         {
-            private static readonly object[] s_empty = new object[0];
+            private static readonly object[] s_empty = Array.Empty<object>();
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactDelegateFactoryBase.cs]: Clean up in separate PR.

**Comment by rozele** [comment]
Please see #1772 for a more minimalist approach.

**Comment by matthargett** [comment]
@reseul any opinions on this one?

**Comment by reseul** [comment]
I don't have anything of value to add, hence my silence :)
This refactoring effort is not something I care that much of currently, so I'm staying aside...



---
In reply to: [382231079](https://github.com/Microsoft/react-native-windows/pull/1768#issuecomment-382231079) [](ancestors = 382231079)

File: ReactWindows/ReactNative.Shared/Bridge/ReactContextNativeModuleBase.cs [hunk 3]
```diff
@@ -72,6 +72,6 @@ protected ReactContextNativeModuleBase(ReactContext reactContext, IReactDelegate
         /// <summary>
         /// The React context.
         /// </summary>
-        public ReactContext Context { get; }
+        public IReactContext Context { get; }
```
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactContextNativeModuleBase.cs]: That's too bad, this is the exact place I'm trying to protect people from abusing. Next time, I guess.

File: ReactWindows/ReactNative.Shared/UIManager/ViewManagerBase.cs [hunk 4]
```diff
@@ -247,14 +248,14 @@ void IViewManager.UpdateProps(object viewToUpdate, JObject props)
...
         {
-            return CreateView(reactContext);
+            return CreateView((ThemedReactContext)reactContext);
         }
 
-        void IViewManager.OnDropViewInstance(ThemedReactContext reactContext, object view)
+        void IViewManager.OnDropViewInstance(IReactContext reactContext, object view)
         {
-            OnDropViewInstance(reactContext, (TView)view);
+            OnDropViewInstance((ThemedReactContext)reactContext, (TView)view);
```
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/UIManager/ViewManagerBase.cs]: literally pulled that from your other PR. it's gross IMO, but I think you said it's a necessary temporary step to phase in a different API. I don't think creating temporary objects each time this is called would be great for performance.

File: ReactWindows/ReactNative/UIManager/BaseViewManager.cs [hunk 5]
```diff
@@ -19,6 +19,7 @@
 using Windows.UI.Xaml.Input;
 using Windows.UI.Xaml.Media;
 using Windows.UI.Xaml.Media.Media3D;
+using ReactNative.Bridge;
```
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative/UIManager/BaseViewManager.cs]: fixed.

File: ReactWindows/ReactNative.Shared/UIManager/IViewManager.cs [hunk 6]
```diff
@@ -5,6 +5,7 @@
 
 using Newtonsoft.Json.Linq;
 using System;
+using ReactNative.Bridge;
```
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/UIManager/IViewManager.cs]: fixed.

File: ReactWindows/ReactNative.Shared/Bridge/ReactContext.cs [hunk 7]
```diff
@@ -462,7 +432,7 @@ public void HandleException(Exception exception)
         /// <remarks>
         /// This method should be called exactly once.
         /// </remarks>
-        internal void InitializeWithInstance(IReactInstance instance)
+        public void InitializeWithInstance(IReactInstance instance)
```
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactContext.cs]: fixed.

File: ReactWindows/ReactNative.Shared/Bridge/IReactContext.cs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+using System;
+
+namespace ReactNative.Bridge
+{
+    /// <inheritdoc />
+    /// <summary>
+    /// Abstract context wrapper for the React instance to manage
+    /// lifecycle events.
+    /// </summary>
+    public interface IReactContext : IAsyncDisposable
```
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/IReactContext.cs]: fixed, but it introduces a type check :/

File: ReactWindows/ReactNative.Shared/Bridge/NativeModuleRegistry.cs [hunk 9]
```diff
@@ -312,10 +312,7 @@ public sealed class Builder
             /// <param name="reactContext">The React context.</param>
             public Builder(ReactContext reactContext)
             {
-                if (reactContext == null)
-                    throw new ArgumentNullException(nameof(reactContext));
-
-                _reactContext = reactContext;
+                _reactContext = reactContext ?? throw new ArgumentNullException(nameof(reactContext));
```
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/NativeModuleRegistry.cs]: fixed.

File: ReactWindows/ReactNative.Shared/Bridge/IReactContext.cs [hunk 10]
```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Portions derived from React Native:
+// Copyright (c) 2015-present, Facebook, Inc.
+// Licensed under the MIT License.
+
+using System;
+
+namespace ReactNative.Bridge
+{
+    /// <inheritdoc />
```
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/IReactContext.cs]: fixed.

File: ReactWindows/ReactNative.Net46/UIManager/BaseViewManager.cs [hunk 11]
```diff
@@ -15,6 +15,7 @@
 using System.Windows.Media;
 using System.Windows.Media.Effects;
 using System.Windows.Media.Media3D;
+using ReactNative.Bridge;
```
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative.Net46/UIManager/BaseViewManager.cs]: fixed.

File: ReactWindows/ReactNative.Shared/UIManager/ViewManagerBase.cs [hunk 12]
```diff
@@ -5,6 +5,7 @@
 using ReactNative.Json;
 using System;
 using System.Collections.Generic;
+using ReactNative.Bridge;
```
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/UIManager/ViewManagerBase.cs]: fixed.

File: ReactWindows/ReactNative.Shared/UIManager/ViewManagerBase.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rozele) [reply ReactWindows/ReactNative.Shared/UIManager/ViewManagerBase.cs]: I think the comment was in reference to your aversion to using the cast :-)

File: ReactWindows/ReactNative.Shared/UIManager/ThemedReactContext.cs [hunk 13]
```diff
@@ -1,75 +1,114 @@
...
+        /// <inheritdoc />
+        public void HandleException(Exception exception)
+        {
+            _reactContext.HandleException(exception);
+        }
+
+        public Task DisposeAsync()
+        {
+            return (_reactContext as IAsyncDisposable)?.DisposeAsync();
+        }
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/UIManager/ThemedReactContext.cs]: We don't need to expose the `DisposeAsync`

File: ReactWindows/ReactNative.Shared/Bridge/ReactContextNativeModuleBase.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rozele) [reply ReactWindows/ReactNative.Shared/Bridge/ReactContextNativeModuleBase.cs]: The point of the PR is to enable an interfaces project, initially. I suspect we'll move to a new namespace for the rest of the modularization, and we can introduce this change at that point.

---
In reply to: [189415092](https://github.com/Microsoft/react-native-windows/pull/1768#discussion_r189415092) [](ancestors = 189415092)

File: ReactWindows/ReactNative.Shared/UIManager/ThemedReactContext.cs (hunk 13 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/UIManager/ThemedReactContext.cs]: are you aware of a way to make it non-public when it's inherited from an interface? I already pushed the interface inheritance down from the IReactContext to the concrete children as part of the previous feedback. I'm not sure what to do here, and I'm pretty sure the original implementation will be necessary later based on my private branch. Let me know how exactly to proceed.

File: ReactWindows/ReactNative.Shared/UIManager/ViewManagerBase.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/UIManager/ViewManagerBase.cs]: fair enough. in a choice between a gross cast versus creating temporary objects that cause unnecessary GC activity, I'll choose the gross cast ;)

File: ReactWindows/ReactNative.Shared/Bridge/ReactContextNativeModuleBase.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactContextNativeModuleBase.cs]: done.

**Pull request closed without merging**